File: src/spirite/base/graphics/DynamicImage.java
package spirite.base.graphics;

import spirite.base.graphics.GraphicsContext.Composite;
import spirite.base.image_data.ImageWorkspace;
import spirite.base.image_data.mediums.ABuiltMediumData.DoerOnGC;
import spirite.base.image_data.mediums.ABuiltMediumData.DoerOnRaw;
import spirite.base.util.MUtil;
import spirite.base.util.linear.MatTrans;
import spirite.base.util.linear.Rect;
import spirite.base.util.linear.MatTrans.NoninvertableException;
import spirite.hybrid.HybridHelper;
import spirite.hybrid.HybridUtil;
import spirite.hybrid.HybridUtil.UnsupportedImageTypeException;

/**
 * DynamicImage
 * 	A wrapper for a RawImage which automatically resizes itself to 
 * 
 *	Although much of this is simply splitting DynamicMedium's code over two objects, since 
 *	other mediums commonly use this behavior, putting it in its own scope.
 */
public class DynamicImage {
	private ImageWorkspace context;
	private int ox, oy;
	private RawImage base;
	private RawImage buffer = null;
	private MatTrans trans = null;
	int checkoutCtr = 0;
	
	public DynamicImage( ImageWorkspace context, RawImage raw, int ox, int oy) {
		this.context = context;
		this.base = raw;
		this.ox = ox;
		this.oy = oy;
	}
	
	public int getXOffset() {return ox;}
	public int getYOffset() {return oy;}
	public int getWidth() {return (base == null) ? 0 : base.getWidth();}
	public int getHeight() {return (base == null) ? 0 : base.getHeight();}
	
	public RawImage getBase() {return base;}
	public ImageWorkspace getContext() {return context;}
	public void setContext( ImageWorkspace ws) {context = ws;}
	
	public void doOnGC( DoerOnGC doer, MatTrans transform) {
		doer.Do(checkoutRaw( transform).getGraphics());
		checkin();
	}
	public void doOnRaw( DoerOnRaw doer, MatTrans transform) {
		doer.Do(checkoutRaw( transform));
		checkin();
	}
	public Rect getDrawBounds( MatTrans transform) {
		return (new Rect( 0, 0, context.getWidth(), context.getHeight())).union(
				MUtil.circumscribeTrans(new Rect(ox, oy, base.getWidth(), base.getHeight()), transform));
	}

	int checkoutCnt = 0;
	private RawImage checkoutRaw(MatTrans transform) {
		if( checkoutCnt++ == 0) {
			this.trans = transform;
			buffer = HybridHelper.createImage(context.getWidth(), context.getHeight());
			GraphicsContext gc = buffer.getGraphics();
			gc.setTransform(trans);
			gc.drawImage(base, ox, oy);
		}
		return buffer;
	}
	
	
	
	private void checkin() {
		if( --checkoutCnt == 0) {
			int w = context.getWidth();
			int h = context.getHeight();
			
			MatTrans invTrans;
	
			try {
				invTrans = trans.createInverse();
			} catch (NoninvertableException e) {
				invTrans = new MatTrans();
			}
			
			Rect drawAreaInImageSpace = MUtil.circumscribeTrans(new Rect(0,0,w,h), invTrans).union(
							new Rect(ox,oy, base.getWidth(), base.getHeight()));
	
			RawImage img = HybridHelper.createImage(drawAreaInImageSpace.width, drawAreaInImageSpace.height);
			GraphicsContext gc = img.getGraphics();
	
			// Draw the old image
			gc.drawImage(base, ox - drawAreaInImageSpace.x, oy - drawAreaInImageSpace.y);
	
			// Clear the section of the old image that will be replaced by the new one
			gc.transform(trans);
			gc.setComposite( Composite.SRC, 1.0f);
			gc.drawImage(buffer, 0, 0);
	//			g2.dispose();
	
			
			Rect cropped = null;
			try {
				cropped = HybridUtil.findContentBounds(img, 0, true);
			} catch (UnsupportedImageTypeException e) {
				e.printStackTrace();
			}
			
			RawImage nri;
			if( cropped == null || cropped.isEmpty()) {
				nri = HybridHelper.createNillImage();
			}
			else {
				nri = HybridHelper.createImage( cropped.width, cropped.height);
			}
			gc = nri.getGraphics();
			gc.drawImage( img, -cropped.x, -cropped.y);
			
			ox = cropped.x-drawAreaInImageSpace.x;
			oy = cropped.y-drawAreaInImageSpace.y;
			
			base.flush();
			base = nri;
			
			buffer.flush();
			img.flush();
			buffer = null;
			gc = null;
		}
	}
	
	public void flush() {
		if( base != null) {
			base.flush();
			base = null;
		}
	}
	
	public DynamicImage deepCopy() {
		DynamicImage copy = new DynamicImage(getContext(), base.deepCopy(), ox, oy);
		return copy;
	}
}

File: src/main/kotlin/spirite/base/graphics/DynamicImage.kt
package spirite.base.graphics

import rb.glow.Composite.SRC
import rb.glow.IFlushable
import rb.glow.img.RawImage
import rb.glow.using
import rb.glow.with
import rb.vectrix.linear.ITransformF
import rb.vectrix.linear.ImmutableTransformF
import rb.vectrix.mathUtil.d
import rbJvm.glow.util.ContentBoundsFinder
import spirite.sguiHybrid.Hybrid
import spirite.base.util.linear.Rect
import spirite.base.util.linear.RectangleUtil


/**
 * DynamicImage
 * 	A wrapper for a RawImage which automatically resizes itself to
 *
 *	Although DynamicImage is essentially a medium (DynamicMedium is a thin wrapper for a DynamicImage), other Mediums and
 *  drawing sources use this dynamically-resizing behavior so it's easiest to abstract it into its own class.
 *
 * Note: There are many ways to do compositing from the composition layers to the base layers.  Can play with them later.
 */
class DynamicImage(
        raw: RawImage? = null,
        xOffset: Int = 0,
        yOffset: Int = 0)
    : IFlushable
{
    var xOffset = xOffset ; private set
    var yOffset = yOffset ; private set
    var base: RawImage? = raw ; private set

    val width get() = base?.width ?: 0
    val height get() = base?.height ?: 0


    fun drawToImage(compositionWidth: Int,
                    compositionHeight: Int,
                    tImageToComposition: ITransformF = ImmutableTransformF.Identity,
                    drawer: (RawImage) -> Unit)
    {
        val checkoutContext = checkoutRaw(compositionWidth, compositionHeight, tImageToComposition)
        drawer.invoke(checkoutContext.buffer)
        checkin(checkoutContext)
    }

    // region Checkin/Checkout
    private inner class CompositionContext(
            val tImageToComposite: ITransformF,
            val compositionWidth : Int,
            val compositionHeight : Int)
    {
        val buffer = Hybrid.imageCreator.createImage(compositionWidth, compositionHeight)
    }

    private fun checkoutRaw(compositionWidth: Int, compositionHeight: Int, tImageToComposition: ITransformF) : CompositionContext {
        val newContext = CompositionContext(tImageToComposition, compositionWidth, compositionHeight)

        val gc = newContext.buffer.graphics
        gc.transform(tImageToComposition)
        val b = base
        if( b != null ) gc.renderImage(b, xOffset.d, yOffset.d)

        return newContext
    }

    private fun checkin(context: CompositionContext) {
        val tCompositeToImage = context.tImageToComposite.invert() ?: ImmutableTransformF.Identity


        // Step 1: Draw Composite and Base to the combining image
        val combiningBounds = RectangleUtil.circumscribeTrans(Rect(0,0, context.compositionWidth, context.compositionHeight), tCompositeToImage)
                .union( Rect(xOffset, yOffset, base?.width ?: 0, base?.height ?: 0))

        using(Hybrid.imageCreator.createImage(combiningBounds.width, combiningBounds.height)) { combiningImage ->
            val gc = combiningImage.graphics

            base?.with {
                gc.renderImage(it, xOffset - combiningBounds.x.d, yOffset - combiningBounds.y.d)
            }

            using(context.buffer) { buffer ->
                gc.transform =tCompositeToImage
                gc.preTranslate(-combiningBounds.x.d, -combiningBounds.y.d)
                gc.composite = SRC
                gc.renderImage(buffer, 0.0, 0.0)

                // Step 2: Crop the Combining Bounds
                val contentBounds = ContentBoundsFinder.findContentBounds(combiningImage, 0, true)
                val newBase = when {
                    contentBounds.w <= 0 || contentBounds.h <= 0 -> null
                    else -> Hybrid.imageCreator.createImage(contentBounds.wi, contentBounds.hi)
                }
                //println("${contentBounds.width} , ${contentBounds.height}")
                newBase?.graphics?.renderImage(combiningImage, -contentBounds.x1, -contentBounds.y1)

                xOffset = contentBounds.x1i + combiningBounds.x
                yOffset = contentBounds.y1i + combiningBounds.y

                base = newBase
            }
        }
    }
    // endregion

    override fun flush() {
        base?.flush()
        base = null
    }
    fun deepCopy() = DynamicImage( base?.deepCopy(), xOffset, yOffset)
}